cmd/gke-certificates-controller/app/gke_approver.go

```diff
@@ -170,34 +189,151 @@ func hasExactUsages(csr *capi.CertificateSigningRequest, usages []capi.KeyUsage)
+		return false
+	}
+	if !hasExactUsages(csr, kubeletServerUsages) {
 		return false
 	}
 	if csr.Spec.Username != x509cr.Subject.CommonName {
 		return false
 	}
 	return true
 }
+
+func validateNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	client := oauth2.NewClient(context.Background(), opts.tokenSource)
+
+	if err := validateNodeServerCertInner(client, opts, csr, x509cr); err != nil {
+		glog.Errorf("validating CSR %q: %v", csr.Name, err)
+		return false
+	}
+	return true
+}
+
+// Only check that DNS name and IP in SAN match an existing VM in the project.
+// Username was already checked against CN, so this CSR is coming from
+// authenticated kubelet.
+func validateNodeServerCertInner(client *http.Client, opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) error {
+	if len(x509cr.DNSNames) == 0 {
+		return errors.New("no SAN DNS names")
+	}
+	if len(x509cr.IPAddresses) == 0 {
+		return errors.New("no SAN IP addresses")
+	}
+	cs, err := compute.New(client)
+	if err != nil {
+		return fmt.Errorf("creating GCE API client: %v", err)
+	}
+
+	var errorMsgs []string
+	for _, n := range x509cr.DNSNames {
+		inst, err := compute.NewInstancesService(cs).Get(opts.projectID, opts.zone, n).Do()
+		if err != nil {
+			errorMsgs = append(errorMsgs, fmt.Sprintf("DNS name %q doesn't match any VM in cluster project/zone: %v", n, err))
+			continue
+		}
+		for _, iface := range inst.NetworkInterfaces {
+			for _, ip := range x509cr.IPAddresses {
+				if ip.String() == iface.NetworkIP {
+					return nil
+				}
+			}
+		}
+		errorMsgs = append(errorMsgs, fmt.Sprintf("IP addresses in CSR (%q) don't match NetworkInterfaces on instance (%v)", x509cr.IPAddresses, inst.NetworkInterfaces))
+	}
+	return errors.New(strings.Join(errorMsgs, "; "))
+}
+
+type approverOptions struct {
+	projectID   string
+	zone        string
+	tokenSource oauth2.TokenSource
+}
+
+func loadApproverOptions(s *GKECertificatesController) (approverOptions, error) {
+	var a approverOptions
+
+	// Get project ID from gce.conf.
+	gceConfig := struct {
+		Global struct {
+			ProjectID string `gcfg:"project-id"`
+		}
+	}{}
+	if err := gcfg.ReadFileInto(&gceConfig, s.GCEConfigPath); err != nil {
+		return a, err
+	}
+	a.projectID = gceConfig.Global.ProjectID
+
+	// Get cluster zone from metadata server.
+	req, err := http.NewRequest(http.MethodGet, "http://metadata.google.internal/computeMetadata/v1/instance/attributes/cluster-location", nil)
+	if err != nil {
+		return a, err
+	}
+	req.Header.Set("Metadata-Flavor", "Google")
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return a, err
+	}
+	defer res.Body.Close()
+	val, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return a, err
+	}
+	a.zone = strings.TrimSpace(string(val))
+
+	// Get the token source for Application Default Credentials.
+	a.tokenSource, err = google.DefaultTokenSource(context.Background())
```

Comment:
Need an AltTokenSource on masters. URLs are in the GCEConfig.

https://github.com/kubernetes/kubernetes/blob/f18a921a0308c4ecd2e7e8b7e30e39a8216a3447/pkg/cloudprovider/providers/gce/token_source.go#L103

Reply:
Done. Is this for quota or permission reasons?